viz_scripts/plots.py

```diff
@@ -48,82 +57,148 @@ def merge_small_entries(labels, values):
+        labels_dist = []
+        values_dist = []
+
+        for x, y in dist_dict.items():
+            labels_dist.append(x)
+            values_dist.append(y)
+
+        return process_trip_data(labels_dist, values_dist)
+    except KeyError:
+        print(f"Column '{df_col}' not found in the data frame.")
+        return pd.DataFrame(), pd.DataFrame()
+    except Exception as e:
+        print(f"An error occurred: {e}")
+        return pd.DataFrame(), pd.DataFrame()
+
+# Create dataframes with cols: 'Label' 'Value' and 'Proportion'
+def process_trip_data(labels, values):
+    """ Inputs:
+    labels = Displayed labels (e.g. "Gas car, drove alone")
+    values = Corresponding vlaues of these labels
+    trip_type = Bar labels (e.g. Labeled by user (Confirmed trips))
+    Returns:
+    df_total_trip_expanded = Data frame without consolidation of Others, used to create the alt_html table
+    df_total_trip = Data frame with consolidation of Others, used to represent the Bar Charts.
+    """
+    if len(labels) == 0 and len(values) == 0:
+        return pd.DataFrame(), pd.DataFrame()
+    m_labels_expanded, m_values_expanded, m_pct_expanded = calculate_pct(labels, values)
+    data_trip_expanded = {'Label': m_labels_expanded, 'Value': m_values_expanded, 'Proportion': m_pct_expanded}
+    df_total_trip_expanded = pd.DataFrame(data_trip_expanded)
+
+    m_labels, m_values, m_pct = merge_small_entries(labels, values)
+    data_trip = {'Label': m_labels, 'Value': m_values, 'Proportion': m_pct}
+    df_total_trip = pd.DataFrame(data_trip)
+    return df_total_trip_expanded, df_total_trip
+
+def plot_and_text_error(e, ax, file_name):
+    stringified_exception = "".join(tb.format_exception(type(e), e, e.__traceback__))
+    ax.text(0,0,s=stringified_exception)
     plt.savefig(SAVE_DIR+file_name+".png", bbox_inches='tight')
+    alt_text = f"Error while generating chart:"
+    alt_text += stringified_exception
+    alt_text = access_alt_text(alt_text, file_name)
+    # TODO: Format the error as HTML instead of plain text
+    alt_html = access_alt_html(alt_text, file_name)
+    return alt_text, alt_html
+
+# Creates/ Appends single bar to the 100% Stacked Bar Chart
+def plot_and_text_stacked_bar_chart(df, df_col, bar_label, ax, text_result, colors, debug_df):
+    """ Inputs:
+    df = Data frame corresponding to the bar in a stacked bar chart
+    bar_name = Text to represent in case data frame is empty (e.g. "Sensed Trip")
+    bar_label = Text to represent the Bar (e.g. Labeled by user\n (Confirmed trips))
+    ax = axis information
+    colors_combined = color mapping dictionary
+    """
+    sns.set(font_scale=1.5)
+    bar_height = 0.2
+    bar_width = [0]
+    try:
+        # TODO: Put this into a dataframe to begin with so that we can use it directly instead of having multiple variables
+        labels = df[df_col].value_counts(dropna=True).keys().tolist()
+        values = df[df_col].value_counts(dropna=True).tolist()
```

Comment:
while this is aggregated by count, just using a different function

We can easily unify these by passing in the aggregation method.
We may also want to consider refactoring further by just writing the single line function to group and aggregate outside the plot function for greater flexibility and to avoid passing functions through to another single line function.

Will finalize this depending on what works best for @Abby-Wheelis

Reply:
Passed in the aggregation method, confirmed that it is working correctly

>    `plot_and_text_stacked_bar_chart(expanded_ct, "Mode_confirm", {distance_col: 'count'}, "Labeled by user\n (Confirmed trips)", ax[0], text_results[0], colors_mode, debug_df)`

gives us

```
            Mode_confirm  distance_miles
0                    Bus              26
1                 E-bike               4
2     E-car, drove alone              35
3     E-car, with others             170
4           Free Shuttle              24
5   Gas Car, drove alone             355
6   Gas Car, with others             266
7             Not a Trip               8
8                  Other              31
9           Regular Bike             112
10         Scooter share               2
11        Taxi/Uber/Lyft               4
12                 Train              33
13                  Walk             347
```

```
 	vals 	pct
Gas Car, drove alone 	355 	25.1
Walk 	347 	24.5
Gas Car, with others 	266 	18.8
E-car, with others 	170 	12.0
Regular Bike 	112 	7.9
E-car, drove alone 	35 	2.5
Train 	33 	2.3
Other 	31 	2.2
Bus 	26 	1.8
Free Shuttle 	24 	1.7
Not a Trip 	8 	0.6
E-bike 	4 	0.3
Taxi/Uber/Lyft 	4 	0.3
Scooter share 	2 	0.1
```

Reply:
Double-checked that the related bar seems to be fine

![image](https://github.com/e-mission/em-public-dashboard/assets/2423263/6082746f-6df9-408e-96c0-00c087a7f0ab)